PIPS/pip-1.md

```diff
@@ -0,0 +1,146 @@
+
+    RegisterAssetTxArgs memory txArgs = RegisterAssetTxArgs({
+        assetHash: assetHash,
+        nativeAssetHash: nativeAssetHash
+    });
+
+    bytes memory txData = _serializeRegisterAssetTxArgs(txArgs);
+    require(eccm.crossChain(nativeChainId, nativeLockProxy, "registerAsset", txData), "EthCrossChainManager crossChain executed error!");
+
+    emit DelegateAsset(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+The `registerAsset` is called by cross-chain transactions from other LockProxies.
+The purpose of this function is to update the `registry` mapping which will be checked when a user calls the `lock` function.
+This helps to ensure that the user does not call `lock` with incorrect parameters.
+
+In the LockProxy, `registerAsset` can be implemented as:
+```
+function registerAsset(bytes memory argsBs, bytes memory fromContractAddr, uint64 fromChainId) onlyManagerContract public {
+  TxRegisterAssetArgs memory args = _deserializTxRegisterAssetArgs(argsBs);
+
+  // `hash` is any supported hashing function
+  key = hash(args.nativeAssetHash, fromChainId, fromContractAddr, args.assetHash);
+
+  require(registry[key] != true);
+  registry[key] = true;
+}
+```
+
+`setManagerProxy` can be restricted to be called once. After it is called the first time, the `managerProxyContract` address cannot be changed.
+
+`bindProxyHash` and `bindAssetHash` functions can be removed.
+
+The LockProxy should have a `lock` function which locks tokens from the user and sends a cross-chain transaction to unlock tokens on the targeted `toChainId` and `targetProxyHash`:
+```
+function lock(address fromAssetHash, uint64 toChainId, bytes memory targetProxyHash, bytes memory toAssetHash, bytes memory toAddress, uint256 amount) public {
+  bytes32 key = hash(fromAssetHash, toChainId, targetProxyHash, toAssetHash);
+  require(registry[key] == true);
+
+  balances[key] += amount;
+
+  // transfer tokens from user to LockProxy
+
+  address fromContractAddr = address(this);
+  TxArgs memory txArgs = TxArgs({
+      fromContractAddr: fromContractAddr,
+      fromAssetHash: fromAssetHash,
+      toAssetHash: toAssetHash,
+      toAddress: toAddress,
+      amount: amount
+  });
+  bytes memory txData = _serializeTxArgs(txArgs);
+
+  require(eccm.crossChain(toChainId, targetProxyHash, "unlock", txData), "EthCrossChainManager crossChain executed error!");
+
+  emit LockEvent(address fromAssetHash, address fromAddress, uint64 toChainId, bytes toAssetHash, bytes toAddress, uint256 amount, ...optional);
+}
+```
+
+The LockProxy should have an `unlock` function which can be called only by the cross-chain manager contract. The function validates the parameters then sends the tokens to the `toAddress`:
+```
+struct TxArgs {
+  bytes fromContractAddr;
+  bytes fromAssetHash;
+  uint64 fromChainId;
+  bytes toAssetHash;
+  bytes toAddress;
+  uint256 amount;
+}
+
+function unlock(bytes memory argsBs) onlyManagerContract {
```

Comment:
The fromChainId and fromContractAddr parameters cannot be removed from unlock method. The reason is the following.

Consider the scenario, you deploy an asset (A_neo) smart contract and a lockproxy (L_neo) in neo blockchain, then you deploy a lockproxy (L_eth) and the correlated asset (A_eth) in ethereum. 

You invoke the lock of L_neo works fine, and later you will receive the A_eth in Ethereum.

Then I'm an attacker, I deploy a smart contract in Neo and construct the TxArgs {fromContractAddr: L_neo, fromAssetHash: A_neo, fromChainId: neoChainId, ...} and do the invocation. The logic of your lockkproxy contract won't identify I'm the attacker.

In current lockproxy contract, the fromContractAddr and fromChainId parameters of unlock are from the unique CrossChainManager.sol contract, which obtains these two value from the poly chain proof.

Reply:
you're right, I'll add the parameters back